### PR TITLE
push open dishwasher

### PR DESCRIPTION
**Pull request by kaviyachandran**

### CODE REVIEW

File: src/giskardpy/goals/align_to_push_door.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                 root_link: str,
+                 tip_link: str,
+                 door_object: str,
+                 # you will need the intermediate point in the direction the object opens. In dishwasher along z axis
+                 # in door along y axis?
+                 door_height: float,
+                 door_length: float,
+                 object_joint_name: str,
+                 tip_gripper_axis: Vector3Stamped,
+                 object_rotation_axis: Vector3Stamped,
```
Comment (ichumuh) [inline src/giskardpy/goals/align_to_push_door.py:31]: the joint has an `axis` parameter

File: src/giskardpy/goals/align_to_push_door.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        root_P_intermediate_point = PointStamped()
+        root_P_intermediate_point.header.frame_id = self.root
+        root_P_intermediate_point.point.x = root_T_door.pose.position.x
+        root_P_intermediate_point.point.y = root_T_door.pose.position.y
+        root_P_intermediate_point.point.z = root_T_door.pose.position.z
+        if root_V_object_rotation_axis.y == 1:
+            # 3/4 of the height as the tip has to be a little further inside the object
+            root_P_intermediate_point.point.z = root_P_intermediate_point.point.z + self.door_height*3/4
+        elif root_V_object_rotation_axis.z == 1:
+            root_P_intermediate_point.point.y = root_P_intermediate_point.point.y + self.door_length*3/4
```
Comment (ichumuh) [inline src/giskardpy/goals/align_to_push_door.py:89]: this assumes that the axis of all doors are aligned the same way?

File: src/giskardpy/goals/align_to_push_door.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                 hold_condition: cas.Expression = cas.FalseSymbol,
+                 end_condition: cas.Expression = cas.TrueSymbol):
+        """
+        The objective is to reach an intermediate point before pushing the door
+        """
+        self.root = god_map.world.search_for_link_name(root_link, root_group)
+        self.tip = god_map.world.search_for_link_name(tip_link, tip_group)
+        self.door_object = god_map.world.search_for_link_name(door_object)
+        self.object_joint_angle = \
+            god_map.world.state.to_position_dict()[god_map.world.search_for_joint_name(object_joint_name)]
```
Comment (ichumuh) [inline src/giskardpy/goals/align_to_push_door.py:48]: `god_map.world.state[joint_name].position`

File: src/giskardpy/goals/align_to_push_door.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        door_rotated_R_door = cas.RotationMatrix.from_axis_angle(axis=object_V_object_rotation_axis,
+                                                                 angle=desired_angle)
+        door_rotated_T_door = cas.TransMatrix(door_rotated_R_door)
+        door_rotated_P_top = cas.dot(door_rotated_T_door, cas.Point3(door_P_intermediate_point))
+        root_P_top = cas.dot(cas.TransMatrix(root_T_door_expr), cas.Point3(door_rotated_P_top))
+
+        minimum_angle_to_push_door = 0.349
+        door_open = cas.greater_equal(self.object_joint_angle, minimum_angle_to_push_door)
+
+        if symbol_manager.evaluate_expr(expr=door_open):
```
Comment (ichumuh) [inline src/giskardpy/goals/align_to_push_door.py:105]: `self.object_joint_angle >= minimum_angle_to_push_door`

File: src/giskardpy/goals/pre_push_door.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        if name is None:
+            name = f'{self.__class__.__name__}/{self.tip}/{self.door_object}'
+
+        super().__init__(name)
+
+        root_T_tip = god_map.world.compose_fk_expression(self.root, self.tip)
+
+        root_P_bottom_left = PointStamped()  # A
+        root_P_bottom_right = PointStamped()  # B
+        root_P_top_left = PointStamped()  # C
```
Comment (ichumuh) [inline src/giskardpy/goals/pre_push_door.py:67]: why don't you work with casadi objects right away?

File: src/giskardpy/goals/pre_push_door.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                 reference_linear_velocity: float = 0.1,
+                 reference_angular_velocity: float = 0.5,
+                 weight: float = WEIGHT_BELOW_CA,
+                 name: Optional[str] = None,
+                 start_condition: cas.Expression = cas.TrueSymbol,
+                 hold_condition: cas.Expression = cas.FalseSymbol,
+                 end_condition: cas.Expression = cas.TrueSymbol):
+        """
+            The objective is to push the object until desired rotation is reached
+        """
```
Comment (ichumuh) [inline src/giskardpy/goals/pre_push_door.py:34]: most of the comments for the first class apply here too

File: test/test_integration_pr2.py [hunk 7]
```diff
@@ -1603,6 +1603,77 @@ def test_open_close_dishwasher(self, kitchen_setup: PR2TestWrapper):
...
+        right_forearm = 'r_forearm'
+        kitchen_setup.register_group(right_forearm,
+                                     root_link_group_name=kitchen_setup.robot_name,
+                                     root_link_name='r_forearm_link')
+        kitchen_setup.set_open_container_goal(tip_link=hand,
+                                              environment_link=handle_name,
+                                              goal_joint_state=1.3217)
+
+        kitchen_setup.allow_collision(group1=door_obj, group2=right_forearm)
+        kitchen_setup.plan_and_execute()
```
Comment (ichumuh) [inline test/test_integration_pr2.py:1675]: this test only checks if the goal is parsed correctly, not if it is executed correctly. can you check automatically, if it's doing the right thing?

**Review by ichumuh — CHANGES_REQUESTED** [review]

File: src/giskardpy/goals/align_to_push_door.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    # start condition - door is open a little bit
+    # hold condition - door is open a little bit
+    # end condition - gripper is in the gap between the open door and the dishwasher
+    def __init__(self,
+                 root_link: str,
+                 tip_link: str,
+                 door_object: str,
+                 # you will need the intermediate point in the direction the object opens. In dishwasher along z axis
+                 # in door along y axis?
+                 door_height: float,
```
[the remark this answers is not in the document]
Comment (kaviyachandran) [reply src/giskardpy/goals/align_to_push_door.py:27]: is it possible to get the dimensions even if the urdf only has a mesh file

File: src/giskardpy/goals/align_to_push_door.py (hunk 2 above)
Comment (kaviyachandran) [reply src/giskardpy/goals/align_to_push_door.py:89]: No. but assumes the door is in yz plane as x would be along object depth

File: src/giskardpy/casadi_wrapper.py [hunk 9]
```diff
@@ -1996,6 +1996,18 @@ def distance_point_to_line(frame_P_point, frame_P_line_point, frame_V_line_direc
...
 
 
+def distance_point_to_plane(frame_P_current, frame_P_bottom_left, frame_P_bottom_right, frame_P_top_left):
+    ab = frame_P_bottom_right - frame_P_bottom_left
+    ac = frame_P_top_left - frame_P_bottom_left
+
+    normal = cross(ab, ac)
+    d = normal.dot(frame_P_bottom_left)
+    normal.scale(1)
+    normal.scale(d)
```
Comment (ichumuh) [inline src/giskardpy/casadi_wrapper.py:2006]: well, there is now point in first scaling it to one and then to d, might as well remove the scale 1

File: src/giskardpy/goals/align_to_push_door.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        root_V_tip_grasp_axis = cas.dot(root_T_tip, tip_V_tip_grasp_axis)
+        root_T_door = god_map.world.compute_fk_pose(self.root, self.door_object)
+        door_T_root = god_map.world.compute_fk_pose(self.door_object, self.root)
+
+        root_P_intermediate_point = cas.Point3(root_T_door.pose.position)
+        if root_V_object_rotation_axis.y == 1:
+            # 3/4 of the height as the tip has to be a little further inside the object
+            root_P_intermediate_point[2] = root_P_intermediate_point[2] + self.door_height*3/4
+        elif root_V_object_rotation_axis.z == 1:
+            root_P_intermediate_point[1] = root_P_intermediate_point[1] + self.door_length*3/4
```
Comment (ichumuh) [inline src/giskardpy/goals/align_to_push_door.py:71]: what if x=1?

File: src/giskardpy/goals/align_to_push_door.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        desired_angle = object_joint_angle * 0.5  # just chose 1/2 of the goal angle
+
+        # find point w.r.t rotated door in local frame
+        door_rotated_R_door = cas.RotationMatrix.from_axis_angle(axis=object_V_object_rotation_axis,
+                                                                 angle=desired_angle)
+        door_rotated_T_door = cas.TransMatrix(door_rotated_R_door)
+        door_rotated_P_top = cas.dot(door_rotated_T_door, door_P_intermediate_point)
+        root_P_top = cas.dot(cas.TransMatrix(root_T_door_expr), door_rotated_P_top)
+
+        minimum_angle_to_push_door = 0.349
```
Comment (ichumuh) [inline src/giskardpy/goals/align_to_push_door.py:84]: hard coding this doesn't seem like the best idea

**Review by ichumuh — CHANGES_REQUESTED** [review]

File: src/giskardpy/goals/align_to_push_door.py (hunk 11 above)
Comment (kaviyachandran) [reply src/giskardpy/goals/align_to_push_door.py:84]: Or I have to make this a parameter

File: src/giskardpy/goals/align_to_push_door.py (hunk 10 above)
Comment (kaviyachandran) [reply src/giskardpy/goals/align_to_push_door.py:71]: that would usually be a drawer that moves along x

**Comment by ichumuh** [comment]
thanks :)

**Pull request merged**